File: apps/ios/Sources/Core/MessageActions/MessageReadPosition.swift
import Foundation

struct ReadSnapshot: Equatable, Sendable { let context: String; let messageIds: [String] }
struct ScrollAnchor: Codable, Equatable, Sendable { let messageId: String; let offset: Int }
@MainActor protocol ScrollAnchorStore {
    func load(_ scope: ActionScope) throws -> ScrollAnchor?
    func save(_ scope: ActionScope, anchor: ScrollAnchor?) throws
}
struct ReadViewToken: Equatable, Sendable { let scope: ActionScope; fileprivate let generation: UInt64 }
final class ReadDisplayPermit: @unchecked Sendable {
    let token: ReadViewToken; let messageId: String; let context: String
    private let lock = NSLock(); private var used = false; private var revoked = false
    fileprivate init(token: ReadViewToken, messageId: String, context: String) { self.token = token; self.messageId = messageId; self.context = context }
    func revoke() { lock.withLock { revoked = true } }
    func claim() throws { try lock.withLock { guard !used, !revoked else { throw MessageActionError.stale }; used = true } }
    func request() throws -> ActionRequest {
        ActionRequest(method: "PUT", path: "rooms/\(token.scope.roomId)/read-state",
            body: try JSONSerialization.data(withJSONObject: ["messageId": messageId, "readContext": context]), successStatus: 200)
    }
}
/// No read reporting on restore. Only a newly visible row after a fresh GET may issue one PUT.
@MainActor final class MessageReadPosition {
    private let anchors: any ScrollAnchorStore
    private var generation: UInt64 = 0
    private var scope: ActionScope?
    private var context: String?
    private(set) var pending: ReadDisplayPermit?
    private(set) var savedMessageIds: [String] = []
    private(set) var needsRefresh = true
    init(anchors: any ScrollAnchorStore) { self.anchors = anchors }
    func select(_ value: ActionScope?) {
        generation &+= 1; pending?.revoke(); pending = nil; scope = value; context = nil; savedMessageIds = []; needsRefresh = true
    }
    func capture() -> ReadViewToken? { scope.map { ReadViewToken(scope: $0, generation: generation) } }
    private func admits(_ token: ReadViewToken) -> Bool { token.scope == scope && token.generation == generation }
    func accept(_ token: ReadViewToken, snapshot: ReadSnapshot) throws -> Bool {
        guard admits(token), pending == nil, context == nil, needsRefresh else { return false }
        guard MessageReadWire.validContext(snapshot.context), snapshot.messageIds.count <= 100, snapshot.messageIds.allSatisfy(actionID) else { throw MessageActionError.invalidResponse }
        context = snapshot.context; savedMessageIds = snapshot.messageIds; needsRefresh = false; return true
    }
    func displayed(_ token: ReadViewToken, messageId: String) throws -> ReadDisplayPermit? {
        guard actionID(messageId) else { throw MessageActionError.invalidResponse }
        guard admits(token), pending == nil, !needsRefresh, let context else { return nil }
        let permit = ReadDisplayPermit(token: token, messageId: messageId, context: context); pending = permit; return permit
    }
    /// For unknown or 409, discard context/queued updates and require GET. Never automatically retry PUT.
    func finish(_ permit: ReadDisplayPermit, acknowledged: Bool, savedMessageId: String?) throws -> Bool {
        guard admits(permit.token), pending === permit else { return false }
        if let savedMessageId, !actionID(savedMessageId) { throw MessageActionError.invalidResponse }
        pending = nil
        if !acknowledged { generation &+= 1; context = nil; needsRefresh = true; savedMessageIds = [] }
        else if let savedMessageId, !savedMessageIds.contains(savedMessageId) { savedMessageIds.append(savedMessageId); savedMessageIds = Array(savedMessageIds.suffix(100)) }
        return true
    }
    func saveAnchor(_ token: ReadViewToken, anchor: ScrollAnchor, currentlyReadable: Set<String>) throws {
        guard actionID(anchor.messageId), anchor.offset >= 0 else { throw MessageActionError.invalidResponse }
        if admits(token), currentlyReadable.contains(anchor.messageId) { try anchors.save(token.scope, anchor: anchor) }
    }
    func restoreAnchor(_ token: ReadViewToken, currentlyReadable: Set<String>) throws -> ScrollAnchor? {
        guard admits(token), let anchor = try anchors.load(token.scope) else { return nil }
        guard actionID(anchor.messageId), anchor.offset >= 0, currentlyReadable.contains(anchor.messageId) else { try anchors.save(token.scope, anchor: nil); return nil }
        return anchor
    }
    func deleted(_ messageId: String) throws {
        guard let scope else { return }
        if try anchors.load(scope)?.messageId == messageId { try anchors.save(scope, anchor: nil) }
        savedMessageIds.removeAll { $0 == messageId }
        if pending?.messageId == messageId { pending?.revoke(); pending = nil; generation &+= 1; context = nil; needsRefresh = true }
    }
    func reset() throws { if let scope { try anchors.save(scope, anchor: nil) }; select(nil) }
}
enum MessageReadWire {
    static func validContext(_ value: String) -> Bool { value.utf8.count == 43 && value.range(of: "^[A-Za-z0-9_-]{42}[AEIMQUYcgkosw048]$", options: .regularExpression) != nil }
    static func get(_ scope: ActionScope) throws -> ActionRequest {
        guard scope.valid else { throw MessageActionError.invalidResponse }
        return ActionRequest(method: "GET", path: "rooms/\(scope.roomId)/read-state", body: nil, successStatus: 200)
    }
    static func snapshot(_ data: Data) throws -> ReadSnapshot {
        let root = try ActionJSON.object(data)
        guard Set(root.keys) == ["readContext", "items"], let context = root["readContext"] as? String, validContext(context),
              let items = root["items"] as? [[String: Any]], items.count <= 100 else { throw MessageActionError.invalidResponse }
        let ids = try items.map { item -> String in
            guard Set(item.keys) == ["messageId"], let id = item["messageId"] as? String, actionID(id) else { throw MessageActionError.invalidResponse }; return id
        }
        return ReadSnapshot(context: context, messageIds: ids)
    }
    static func saved(_ data: Data) throws -> String? {
        let root = try ActionJSON.object(data)
        guard Set(root.keys) == ["messageId"] else { throw MessageActionError.invalidResponse }
        if root["messageId"] is NSNull { return nil }
        guard let id = root["messageId"] as? String, actionID(id) else { throw MessageActionError.invalidResponse }; return id
    }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/messageactions/MessageReadPosition.kt
package chat.rogi.rogichat.core.messageactions

import chat.rogi.rogichat.core.auth.StrictAuthJson
import kotlinx.serialization.json.*
import chat.rogi.rogichat.core.network.M11Dtos
import chat.rogi.rogichat.core.network.ReadContext
import chat.rogi.rogichat.core.network.ReadStateId

/** Own display progress is not unread count, delivery ACK, a stream order or a sync cursor. */
data class ReadSnapshot(val context: String, val messageIds: List<String>, val firstUnreadMessageId: String? = null)
data class ScrollAnchor(val messageId: String, val offset: Int) {
    init { actionId(messageId); require(offset >= 0) }
}
interface ScrollAnchorStore {
    fun load(scope: ActionScope): ScrollAnchor?
    fun save(scope: ActionScope, anchor: ScrollAnchor?)
}
class ReadViewToken internal constructor(val scope: ActionScope, internal val generation: Long) {
    override fun equals(other: Any?) = other is ReadViewToken && scope == other.scope && generation == other.generation
    override fun hashCode() = 31 * scope.hashCode() + generation.hashCode()
}
class ReadDisplayPermit internal constructor(val token: ReadViewToken, val messageId: String,
    val context: String, private val owner: MessageReadPosition) {
    private var used = false
    fun claim() = synchronized(owner) {
        check(!used && owner.admits(token) && owner.pending === this) { "stale_read_context" }; used = true
    }
    fun request() = ActionRequest("PUT", "rooms/${token.scope.roomId}/read-state",
        M11Dtos.displayed(ReadStateId(messageId), ReadContext(context)), 200)
}
class MessageReadPosition(private val anchors: ScrollAnchorStore) {
    private var generation = 0L
    private var scope: ActionScope? = null
    private var context: String? = null
    var pending: ReadDisplayPermit? = null; private set
    var savedMessageIds: List<String> = emptyList(); private set
    var needsRefresh = true; private set
    @Synchronized fun select(value: ActionScope?) {
        generation++; scope = value; context = null; pending = null; savedMessageIds = emptyList(); needsRefresh = true
    }
    @Synchronized fun capture() = scope?.let { ReadViewToken(it, generation) }
    internal fun admits(token: ReadViewToken) = scope == token.scope && generation == token.generation
    @Synchronized fun accept(token: ReadViewToken, snapshot: ReadSnapshot): Boolean {
        if (!admits(token) || pending != null || context != null || !needsRefresh) return false
        ReadContext(snapshot.context); require(snapshot.messageIds.size <= 100); snapshot.messageIds.forEach(::actionId)
        context = snapshot.context; savedMessageIds = snapshot.messageIds; needsRefresh = false; return true
    }
    /** Call ONLY for a newly visible authorized row after context GET; never replay restored positions. */
    @Synchronized fun displayed(token: ReadViewToken, messageId: String): ReadDisplayPermit? {
        actionId(messageId)
        if (!admits(token) || pending != null || needsRefresh || messageId in savedMessageIds) return null
        return ReadDisplayPermit(token, messageId, context ?: return null, this).also { pending = it }
    }
    /** Unknown/409 invalidates the context and drops queued display events; no automatic PUT retry. */
    @Synchronized fun finish(permit: ReadDisplayPermit, acknowledged: Boolean, savedMessageId: String?): Boolean {
        if (!admits(permit.token) || pending !== permit) return false
        savedMessageId?.let(::actionId); pending = null
        if (!acknowledged) { generation++; context = null; needsRefresh = true; savedMessageIds = emptyList() }
        else if (savedMessageId != null) savedMessageIds = (savedMessageIds + savedMessageId).distinct().takeLast(100)
        return true
    }
    @Synchronized fun saveAnchor(token: ReadViewToken, anchor: ScrollAnchor, currentlyReadable: Set<String>) {
        if (admits(token) && anchor.messageId in currentlyReadable) anchors.save(token.scope, anchor)
    }
    @Synchronized fun restoreAnchor(token: ReadViewToken, currentlyReadable: Set<String>): ScrollAnchor? {
        if (!admits(token)) return null
        val anchor = anchors.load(token.scope) ?: return null
        if (anchor.messageId !in currentlyReadable) { anchors.save(token.scope, null); return null }
        return anchor
    }
    @Synchronized fun deleted(messageId: String) {
        val current = scope ?: return
        if (anchors.load(current)?.messageId == messageId) anchors.save(current, null)
        savedMessageIds = savedMessageIds.filterNot { it == messageId }
        if (pending?.messageId == messageId) { generation++; pending = null; context = null; needsRefresh = true }
    }
    @Synchronized fun reset() { scope?.let { anchors.save(it, null) }; select(null) }
}
object MessageReadWire {
    fun get(scope: ActionScope) = ActionRequest("GET", "rooms/${scope.roomId}/read-state", null, 200)
    fun snapshot(body: String): ReadSnapshot {
        val root = StrictAuthJson.objectValue(body)
        require(root.keys == setOf("readContext", "items") ||
            root.keys == setOf("readContext", "items", "firstUnreadMessageId"))
        require(root.getValue("items").jsonArray.all { it.jsonObject.keys == setOf("messageId") })
        val value = M11Dtos.readStates(body)
        // Public GET projection only contains authorized IDs, unlike nullable PUT receipt.
        require(value.items.all { it.messageId != null })
        return ReadSnapshot(value.readContext.value, value.items.map { it.messageId!!.value }, value.firstUnreadMessageId?.value)
    }
    fun saved(body: String): String? {
        require(StrictAuthJson.objectValue(body).keys == setOf("messageId"))
        return M11Dtos.readState(body).messageId?.value
    }
}
